core/src/main/java/com/linecorp/armeria/server/AnnotatedHttpService.java

```diff
@@ -69,15 +72,34 @@ boolean overlaps(AnnotatedHttpService entry) {
 
     @Override
     public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exception {
-        final Object ret = function.apply(ctx, req);
+
+        final Object ret;
+        try {
+            ret = function.apply(ctx, req);
+        } catch (IllegalArgumentException ignore) {
+            throw new HttpResponseException(HttpStatus.BAD_REQUEST);
+        }
+
         if (!(ret instanceof CompletionStage)) {
             return HttpResponse.ofFailure(new IllegalStateException(
                     "illegal return type: " + ret.getClass().getSimpleName()));
         }
 
         @SuppressWarnings("unchecked")
         CompletionStage<HttpResponse> castStage = (CompletionStage<HttpResponse>) ret;
-        return HttpResponse.from(castStage);
+        return HttpResponse.from(castStage.handle((httpResponse, throwable) -> {
+            if (throwable != null) {
+                if (throwable instanceof CompletionException &&
```

Comment:
I'm curious should I use Guava's `getRootCause` or just check 1 depth of cause.

Reply:
That's a good question. +1 for using Guava.

Reply:
Let me merge once you switch to Guava's `getRootCause()`